### PR TITLE
Create and use OpticalFlow singleton

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: ArduPlane/ArduPlane.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ const AP_Scheduler::Task Plane::scheduler_tasks[] = {
     SCHED_TASK(compass_cal_update,     50,    50),
     SCHED_TASK(accel_cal_update,       10,    50),
 #if OPTFLOW == ENABLED
-    SCHED_TASK(update_optical_flow,    50,    50),
+    SCHED_TASK_CLASS(OpticalFlow, &plane.optflow, update,    50,    50),
```
Comment (WickedShell) [inline ArduPlane/ArduPlane.cpp]: Not to scope creep it, but I notice that copter runs optical flow at 200Hz. Given that quadplanes might want to be using optical flow as well should we be raising this to a 200Hz task on plane as well?

File: libraries/AP_OpticalFlow/OpticalFlow.cpp [hunk 2]
```diff
@@ -120,10 +118,66 @@ void OpticalFlow::init(void)
...
+    Log_Write_Optflow();
+}
+
+void OpticalFlow::Log_Write_Optflow()
+{
+    DataFlash_Class *instance = DataFlash_Class::instance();
+    if (instance == nullptr) {
+        return;
+    }
+    if (_log_bit == (uint32_t)-1 ||
```
Comment (WickedShell) [inline libraries/AP_OpticalFlow/OpticalFlow.cpp]: Doesn't this cause plane to always log while disarmed regardless of what the user wants since the bit is -1?

Would it be reasonable to extend `DataFlash_Class::should_log(const uint32_t mask)` to accept/manage the special case of -1? (Alternatively you could just pass in -1 anyways and that would mean if any logging is enabled that we would log this message (which seems appropriate since you are trying to always log anyways)).

Comment (peterbarker) [reply libraries/AP_OpticalFlow/OpticalFlow.cpp]: So I've changed behaviour here for Copter and Sub but not Plane.  Previously they all would have ignored the arm and download states etc etc.  Now only Plane does....

This bit bothers me a lot, actually.  AP_InertialSensor_Backend also takes a `-1` bit for logging - but that value over there means "never log raw imu data".  This makes it problematic to handle this in the DataFlash library.

And what's "Resolve conversation"?! :-)

Comment (WickedShell) [reply libraries/AP_OpticalFlow/OpticalFlow.cpp]: Means we can dismiss a point I raised as "acceptable" without you having to change code, and makes it easier for future reviewers. :)

File: ArduPlane/ArduPlane.cpp (hunk 1 above)
Comment (WickedShell) [reply ArduPlane/ArduPlane.cpp]: Fair enough. I don't have the hardware so I can't volunteer...

File: libraries/AP_OpticalFlow/OpticalFlow.cpp [hunk 3]
```diff
@@ -127,3 +129,14 @@ void OpticalFlow::update(void)
...
 }
 
+// singleton instance
+OpticalFlow *OpticalFlow::_singleton;
+
+namespace AP {
+
+OpticalFlow *opticalflow()
+{
+    return OpticalFlow::get_singleton();
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/OpticalFlow.cpp]: can we normalize on the name? Some times we have `get_instance()`, some times `get_singleton()`, some times `instance()`

File: libraries/AP_OpticalFlow/OpticalFlow.h [hunk 4]
```diff
@@ -36,6 +36,11 @@ class OpticalFlow
     OpticalFlow(const OpticalFlow &other) = delete;
     OpticalFlow &operator=(const OpticalFlow&) = delete;
 
+    // get singleton instance
+    static OpticalFlow *get_singleton() {
+        return _singleton;
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/OpticalFlow.h]: Just a note that this is not really a singleton pattern... For a real singleton you'd have:

OpticalFlow.cpp:
```
OpticalFlow OpticalFlow::_singleton;
```

Then the method would return `&singleton`.

The difference is that with this approach we are implicitly forcing the object to be constructed when someone tries to get it. In your current approach, a NULL pointer can still be returned if the object wasn't constructed by other means when this function is called.

File: ArduCopter/GCS_Mavlink.cpp [hunk 5]
```diff
@@ -329,13 +329,6 @@ bool GCS_MAVLINK_Copter::try_send_message(enum ap_message id)
 #endif // MOUNT == ENABLED
         break;
 
-    case MSG_OPTICAL_FLOW:
-#if OPTFLOW == ENABLED
-        CHECK_PAYLOAD_SIZE(OPTICAL_FLOW);
-        send_opticalflow(copter.optflow);
-#endif
-        break;
-
```
Comment (lucasdemarchi) [inline ArduCopter/GCS_Mavlink.cpp]: sigh about convention of splitting per-dir changes :( - "move", but where's the rest?

**Comment by peterbarker** [comment]
On Tue, 4 Sep 2018, Lucas De Marchi wrote:

> +    return OpticalFlow::get_singleton();
> 
> can we normalize on the name? Some times we have get_instance(), some times
> get_singleton(), some times instance()

I believe we've decided to canonicalise on get_singleton, to avoid 
confusion with the term "instance", often used to refer to the backend 
"instances".

I'll do a sweep through the code at some stage to canonicalise them, I 
think - assuming we can reach consensus on naming!

**Comment by peterbarker** [comment]
On Tue, 4 Sep 2018, Lucas De Marchi wrote:

> sigh about convention of splitting per-dir changes :( - "move", but where's
> the rest?

That convention was like that when I got here! :-)

File: libraries/AP_OpticalFlow/OpticalFlow.cpp (hunk 3 above)
Comment (WickedShell) [reply libraries/AP_OpticalFlow/OpticalFlow.cpp]: I have a strong preference for `get_singleton()` or `singleton()`. Instance is very overloaded in any of the drivers that have backends (IE GPS/airspeed/compass)

**Comment by peterbarker** [comment]
I haven't flown this, but both Plane and Copter both log to dataflash and send mavlink OPTICAL_FLOW.

**Comment by peterbarker** [comment]
Sub logs to dataflash but doesn't send any mavlink messages; the "healthy" call returns false.  It also does that on master so *shrug*.
